File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.py
from collections import deque

T = int(input())
dx, dy = [-2, -1, 1, 2, 2, 1, -1, -2], [1, 2, 2, 1, -1, -2, -2, -1]

def bfs(s_x, s_y, t_x, t_y):
    q = deque([(s_x, s_y)])
    visited[s_x][s_y] = 0

    while q:
        x, y = q.popleft()
        if x==t_x and y==t_y:
            return visited[x][y]

        for i in range(8):
            nx, ny = x+dx[i], y+dy[i]
            if 0<=nx<I and 0<=ny<I and visited[nx][ny]==-1:
                q.append((nx, ny))
                visited[nx][ny] = visited[x][y] + 1


for _ in range(T):
    I = int(input())
    visited = [[-1]*I for _ in range(I)]
    start_x, start_y = map(int, input().split())
    target_x, target_y = map(int, input().split())

    print(bfs(start_x, start_y, target_x, target_y))

File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>

using namespace std;

int dx[8] = {1, 2, 2, 1, -1, -2, -2, -1};
int dy[8] = {2, 1, -1, -2, -2, -1, 1, 2};

int bfs(int I, int x, int y, int target_x, int target_y){
    queue<pair<int, int>> q;
    vector<vector<int>> visited(I, vector<int>(I, -1));
    q.push({x, y});
    visited[x][y] = 0;

    while(!q.empty()){
        auto[cx, cy] = q.front();
        q.pop();
        if(cx==target_x && cy==target_y){
            return visited[cx][cy];
        }

        for(int i=0; i<8; i++){
            int nx = cx + dx[i];
            int ny = cy + dy[i];
            if(0<=nx && nx<I && 0<=ny && ny<I && visited[nx][ny]==-1){
                q.push({nx, ny});
                visited[nx][ny] = visited[cx][cy] + 1;
            }
        }
    }
    return -1;
}

int main(void){
    int T;
    cin >> T;
    while(T--){
        int I;
        cin >> I;
        int cx, cy;
        cin >> cx >> cy;
        int target_x, target_y;
        cin >> target_x >> target_y;

        int result = bfs(I, cx, cy, target_x, target_y);

        cout << result << endl;
    }


    return 0;
}
